bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java

```diff
@@ -0,0 +1,244 @@
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.cache.ExpiringCache;
+import org.eclipse.smarthome.core.i18n.LocationProvider;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.PointType;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.fmiweather.internal.BindingConstants;
+import org.openhab.binding.fmiweather.internal.client.Client;
+import org.openhab.binding.fmiweather.internal.client.Location;
+import org.openhab.binding.fmiweather.internal.client.exception.FMIResponseException;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link FMIDiscoveryService} creates things based on the configured location.
+ *
+ * @author Sami Salonen - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.fmiweather")
+public class FMIWeatherDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(FMIWeatherDiscoveryService.class);
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_OBSERVATION);
+    private static final long STATIONS_CACHE_MILLIS = TimeUnit.HOURS.toMillis(12);
+    private static final int STATIONS_TIMEOUT_MILLIS = (int) TimeUnit.SECONDS.toMillis(10);
+    private static final int DISCOVER_TIMEOUT_SECONDS = 5;
+    private static final int LOCATION_CHANGED_CHECK_INTERVAL_SECONDS = 60;
+    private static final int FIND_STATION_METERS = 80_000;
+
+    private @Nullable ScheduledFuture<?> discoveryJob;
+    private @Nullable PointType previousLocation;
+    private @NonNullByDefault({}) LocationProvider locationProvider;
+
+    private ExpiringCache<Set<Location>> stationsCache = new ExpiringCache<>(STATIONS_CACHE_MILLIS, () -> {
+        try {
+            return new Client().queryWeatherStations(STATIONS_TIMEOUT_MILLIS);
+        } catch (FMIResponseException e) {
+            logger.warn("Error when querying stations", e);
```

Comment:
Would logging the exception's message (instead of the stack trace) sufficient here?

Reply:
Does that log `cause` of the exception? I think that will be important.

Reply:
Logging the stack trace when the HTTP request times out seems a bit unreasonable. If the cause is important here, why not logging it together with the message?

Reply:
yes, I was thinking more on the SAX exceptions etc. but good point that timeouts would be logged quite verbosely now.

I'll see what I can do to make it nicer.

Reply:
All right, now logging only unexpected-content errors (e.g xml errors) verbosely : 6c65be6bb1